src/compiler/parser.ts

```diff
@@ -6601,6 +6611,25 @@ namespace Parser {
             case SyntaxKind.NewKeyword:
                 return parseNewExpressionOrNewDotTarget();
             case SyntaxKind.SlashToken:
+                // JSX inside a .ts (not .tsx) file gets a specialized error...

```

Comment:
From https://github.com/microsoft/TypeScript/pull/56101#pullrequestreview-1735462101:

> I almost wish we could add this message to an existing error that appears with JSX, only in cases where we can be sure that the cause really is a tag. Would that be a better approach? I'm not sure any of the existing errors have enough context to decide that they were really, truly caused by a JSX tag.

I looked into that at first but couldn't find any good path forward 😞. JSX syntax issues tend to cause multiple seemingly unrelated parsing errors. If you have a good path I would be _very_ happy to go down it.

Reply:
There should be a TypeAssertion somewhere earlier in the parser tree -- but we don't know something has gone wrong until much later, when we see an unexpected `/` that doesn't end up parsing as a regex.

What if you are extra suspicious of TypeAssertionExpressions? Maybe after parsePrimaryExpression you can do a speculative check for illegal stuff. That should at least let you detect small things like `<x>y</x>`, and probably also illegal things like `<x>y z</x>`. In theory that would then let you do some error recovery to skip past the closing tag, but that sounds quite difficult.